src/orbit-common/key-map.js

```diff
@@ -0,0 +1,95 @@
+import { get } from 'orbit/lib/objects';
+import { firstResult } from 'orbit/lib/arrays';
+import { assert } from 'orbit/lib/assert';
+
+export default class KeyMap {
+  constructor() {
+    this._data = {};
+  }
+
+  idToKey(type, keyName, idValue, autoGenerate) {
+    // TODO: find instances of using autoGenerate and refactor
+    assert('stop using autogenerate', !autoGenerate);
+    return get(this._data, type, keyName, 'idToKeyMap', idValue);
+  }
+
+  keyToId(type, keyName, keyValue, autoGenerate) {
+    // TODO: find instances of using autoGenerate and refactor
+    assert('stop using autogenerate', !autoGenerate);
+    return get(this._data, type, keyName, 'keyToIdMap', keyValue);
+  }
+
+  /**
+   Given a data object structured according to this schema, register all of its
+   key mappings. This data object may contain any number of records and types.
+
+   @param {Object} data - data structured according to this schema
+   @returns {undefined}
+   */
+  registerDocument(data) {
+    if (!data) {
+      return;
+    }
+
+    Object.keys(data).forEach(type => {
+      let idRecordMap = data[type];
+      Object.keys(idRecordMap).forEach(id => {
+        let record = idRecordMap[id];
+        this.push({
+          type,
+          id: record.id,
+          keys: record.keys
+        });
+      });
+    });
+  }
+
+  // TODO: use _.set pattern to clean this up and accept one key at a time per record
+  push({ type, id, keys }) {
```

Comment:
Perhaps `register` or `registerRecord` instead of `push` to be consistent with `registerDocument` ?